src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/AbstractDocumentationCommentService.cs

```diff
@@ -0,0 +1,176 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using Roslyn.Utilities;
+using System.Text;
+
+namespace Microsoft.CodeAnalysis.LanguageServices
+{
+    internal abstract class AbstractDocumentationCommentService<
+        TDocumentationCommentTriviaSyntax,
+        TXmlNodeSyntax,
+        TXmlAttributeSyntax,
+        TCrefSyntax,
+        TXmlElementSyntax,
+        TXmlTextSyntax,
+        TXmlEmptyElementSyntax,
+        TXmlCrefAttributeSyntax,
+        TXmlNameAttributeSyntax,
+        TXmlTextAttributeSyntax> : IDocumentationCommentService
+        where TDocumentationCommentTriviaSyntax : SyntaxNode
+        where TXmlNodeSyntax : SyntaxNode
+        where TXmlAttributeSyntax : SyntaxNode
+        where TCrefSyntax : SyntaxNode
+        where TXmlElementSyntax : TXmlNodeSyntax
+        where TXmlTextSyntax : TXmlNodeSyntax
+        where TXmlEmptyElementSyntax : TXmlNodeSyntax
+        where TXmlCrefAttributeSyntax : TXmlAttributeSyntax
+        where TXmlNameAttributeSyntax : TXmlAttributeSyntax
+        where TXmlTextAttributeSyntax : TXmlAttributeSyntax
+    {
+        public const string Ellipsis = "...";
+        public const int MaxXmlDocCommentBannerLength = 120;
+
+        private readonly ISyntaxFactsService _syntaxFacts;
+
+        protected AbstractDocumentationCommentService(ISyntaxFactsService syntaxFacts)
+        {
+            _syntaxFacts = syntaxFacts;
+        }
+
+        private static void AddSpaceIfNotAlreadyThere(StringBuilder sb)
+        {
+            if (sb.Length > 0 && sb[sb.Length - 1] != ' ')
+            {
+                sb.Append(' ');
+            }
+        }
+
+        private string GetDocumentationCommentPrefix(TDocumentationCommentTriviaSyntax documentationComment)
+        {
+            Contract.ThrowIfNull(documentationComment);
+
+            var leadingTrivia = documentationComment.GetLeadingTrivia();
+            var exteriorTrivia = leadingTrivia.Where(t => _syntaxFacts.IsDocumentationCommentExteriorTrivia(t))
+                                              .FirstOrNullable();
+
+            return exteriorTrivia != null ? exteriorTrivia.Value.ToString() : string.Empty;
+        }
+
+        public string GetBannerText(TDocumentationCommentTriviaSyntax documentationComment, CancellationToken cancellationToken)
+        {
+            // TODO: Consider unifying code to extract text from an Xml Documentation Comment (https://github.com/dotnet/roslyn/issues/2290)
```

Comment:
I tried, and it was non trivial.  The issue is that thisis called through synchronous paths that don't have access to a semantic model.  I do *not* want to synchronously try to get a semantic model  to support that.

One thing i considerd was that we could have one code path for teh synchronous load codepath.  But then if it's the async codepath we could do the other path.

This could be weird though because you could observe the banner text changing in the buffer on open.  i.e. a a cref would be displayed one way initially, then would change as we computed the semantic version.

Effectively there were too many open issues and i didn't want to handle that yet.  Is that ok?